nucypher/blockchain/economics.py

```diff
@@ -14,60 +14,167 @@
-            self,
-            initial_supply=10**9,
-            initial_inflation=1.0,
-            T_half=2.0,
-            T_sat=1.0,
-            small_staker_multiplier=0.5) -> None:
+    # Value Constraints
+    minimum_allowed_locked = NU(15_000, 'NU').to_nunits()
+    maximum_allowed_locked = NU(4_000_000, 'NU').to_nunits()
+
+    # Supply
+    __default_initial_supply = NU(int(1_000_000_000), 'NU').to_nunits()
+    __default_initial_inflation = 1
+    __default_token_halving = 2      # years
+    __default_reward_saturation = 1  # years
+    __default_small_stake_multiplier = Decimal(0.5)
+
+    def __init__(self,
+                 initial_supply: int = __default_initial_supply,
+                 initial_inflation: int = __default_initial_inflation,
+                 halving_delay: int = __default_token_halving,
+                 reward_saturation: int = __default_reward_saturation,
+                 small_stake_multiplier: Decimal = __default_small_stake_multiplier
+                 ):
         """
-        Calculate parameters to use in token issuer from high-level
-        human-understandable parameters
-
-        :param initial_supply: coins at t=0,
-        :param initial_inflation: inflation in day 1 expressed in units of year**-1
-            e.g. 1/365 of all coins in day 1 is 1 year**-1
-        :param T_half: time for inflation halving in years
-        :param T_sat: "saturation" time - if staking is longer than T_sat, the
-            reward doesn't get higher
-        :param small_staker_multiplier: what fraction of max reward rate do we
-            pay to those who are about to unlock coins
+        :param initial_supply: Tokens at t=0
+        :param initial_inflation: Inflation on day 1 expressed in units of year**-1
+        :param halving_delay: Time for inflation halving in years
+        :param reward_saturation: "saturation" time - if staking is longer than T_sat, the reward doesn't get any higher
+        :param small_stake_multiplier: Fraction of maximum reward rate paid to those who are about to unlock tokens
         """
 
-        # Saving input in case we'll need it
-        self.input_params = dict(
-                initial_supply=initial_supply,
-                initial_inflation=initial_inflation,
-                T_half=T_half,
-                T_sat=T_sat,
-                small_staker_multiplier=small_staker_multiplier)
+        #
+        # Calculate
+        #
+
+        with localcontext() as ctx:
+            ctx.prec = self._precision
+
+            initial_supply = Decimal(initial_supply)
+
+            # ERC20 Token parameter
+            total_supply = initial_supply * (1 + initial_inflation * halving_delay / LOG2)
+
+            # Remaining / Reward Supply - Escrow Parameter
+            reward_supply = total_supply - initial_supply
+
+            # k2 - Escrow parameter
+            staking_coefficient = 365 ** 2 * reward_saturation * halving_delay / LOG2 / (1 - small_stake_multiplier)
```

Comment:
This is called `mining_coefficient` in `Issuer.sol`.

Reply:
I think I like `staking_coefficient` more, but I want to remain conventional with the reference. What do you think?

Reply:
It's not that this name is frozen in the smart contract, can be changed there also. We just need to decide which name to stick with.
Either are fine with me btw

Reply:
Is there an issue open (or even a resolution) to this re: going from `mining_coefficient` -> `staking_coefficient`

Reply:
Why is the `K2` coefficient called `mining_coefficient`? After following the same procedure that @michwill devised, I understand its origin, but I don't get at all why this is a "mining coefficient". How is it related to mining or staking? As far as I can tell, it's an arbitrary change of variable. Is that's the case, then the name is also arbitrary.

Reply:
This is also applicable to the "Locked periods coefficient" (K1)

Reply:
`mining_coefficient` is an arbitrary name!